core/chain_indexer.go

```diff
@@ -0,0 +1,296 @@
+	Process(header *types.Header)   // process a single block (called for each block in the section)
+	Commit(db ethdb.Database) error // do some more processing if necessary and store the results in the database
+	UpdateMsg(done, all uint64)     // print a progress update message if necessary (only called when multiple sections need to be processed)
+}
+
+// NewChainIndexer creates a new  ChainIndexer
+//  db:				database where the index of available processed sections is stored (the index is stored by the
+//                  indexer, the actual processed chain data is stored by the backend)
+//  dbKey:			key prefix where the index is stored
+//  backend:		an implementation of ChainIndex
+//  sectionSize:	the size of processable sections
+//  confirmReq:		required number of confirmation blocks before a new section is being processed
+//  procWait:		waiting time between processing sections (simple way of limiting the resource usage of a db upgrade)
+//  stop:		    quit channel
+func NewChainIndexer(db ethdb.Database, dbKey []byte, backend ChainIndex, sectionSize, confirmReq uint64, procWait time.Duration, stop chan struct{}) *ChainIndexer {
+	c := &ChainIndexer{
+		db:                   db,
+		validSectionsKey:     append(dbKey, []byte("-count")...),
+		sectionHeadKeyPrefix: append(dbKey, []byte("-shead")...),
+		backend:              backend,
+		sectionSize:          sectionSize,
+		confirmReq:           confirmReq,
+		tryUpdate:            make(chan struct{}, 1),
+		stop:                 stop,
+		procWait:             procWait,
+	}
+	c.stored = c.getValidSections()
+	go c.updateLoop()
+	return c
+}
+
+// updateLoop is the main event loop of the indexer
+func (c *ChainIndexer) updateLoop() {
+	updateMsg := false
+
+	for {
+		select {
+		case <-c.stop:
+			return
+		case <-c.tryUpdate:
+			c.lock.Lock()
+			if c.targetCount > c.stored {
+				if !updateMsg && c.targetCount > c.stored+1 {
+					updateMsg = true
+					c.backend.UpdateMsg(c.stored, c.targetCount)
+				}
+				c.calcIdx = c.stored
+
+				var lastSectionHead common.Hash
+				if c.calcIdx > 0 {
+					lastSectionHead = c.getSectionHead(c.calcIdx - 1)
+				}
+
+				c.lock.Unlock()
+				sectionHead, ok := c.processSection(c.calcIdx, lastSectionHead)
+				c.lock.Lock()
+
+				if ok && lastSectionHead == c.getSectionHead(c.calcIdx-1) {
+					c.stored = c.calcIdx + 1
+					c.setSectionHead(c.calcIdx, sectionHead)
+					c.setValidSections(c.stored)
+					if updateMsg {
+						c.backend.UpdateMsg(c.stored, c.targetCount)
+						if c.stored >= c.targetCount {
+							updateMsg = false
+						}
+					}
+					c.lastForwarded = c.stored*c.sectionSize - 1
+					for _, cp := range c.children {
+						cp.newHead(c.lastForwarded, false)
+					}
+				} else {
+					// if processing has failed, do not retry until further notification
+					c.targetCount = c.stored
+				}
+			}
+
+			if c.targetCount > c.stored {
+				go func() {
+					time.Sleep(c.procWait)
```

Comment:
It makes sense when generating an index for a large number of sections. It is a simple but effective way of ensuring that it will leave some processing power for the other tasks too.